Added cpupower test cases.

PavithraDP opened the pull request:
The test case does the following:
1)To get the initial frequency and governor set on system
2)Get the list of governors available on the system
3)Get the cur,min,max freequency on the system
4)Validate Performance,Powersave,Ondemand Governor
5)Set the final values which was taken initially.

PraveenPenguin commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+
+    def test(self):
+        self.nfail = 0
+        self.i = 0
```
please use a meaningful name of variable

PraveenPenguin commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+            if (gov == "userspace"):
+                self.log.info("Check for userspace governor")
+                self.check_userspace_governor(gov)
+            if (gov == "ondemand" or "conservative" or "schedutils"):
```
can it be` if gov in [ "ondemand" ,"conservative" , "schedutils"]` ?

PraveenPenguin replied:
can it be` if gov in [ "ondemand" ,"conservative" , "schedutils"]` ?

sathnaga replied:
it would be good to have a set_governor() and check_governor() as generic functions so that lot of if else can be avoided

PavithraDP replied:
done

PraveenPenguin commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        cmd = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_cur_freq" % (self.i)
+        initial_freq = process.system_output(cmd)
+        cmd1 = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_governor" % (self.i)
+        initial_governor = process.system_output(cmd1)
```
can you please use same variable as `cmd` instead of `cmd1` as using extra variable increase  python garbage collector work :)

PraveenPenguin replied:
can you please use same variable as `cmd` instead of `cmd1` as using extra variable increase  python garbage collector work :)

PavithraDP replied:
done

PraveenPenguin commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        """
+        Get the list of governors available on the system
+        """
+        cmd_governor_list = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_available_governors" % (self.i)
```
here also `cmd` can be used

PavithraDP replied:
done

apahim replied:
I'd prefer to see a python `open().read()` here instead of a `process.system_output('cat ...')` (and everywhere else).

PavithraDP replied:
Have handled fail condition with error_count instead of "True" or "False"
Included changes to read file using open().read()

PraveenPenguin commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        cmd_governor_list = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_available_governors" % (self.i)
+        output = process.system_output(cmd_governor_list)
+        governors = output.split()
+        return governors
```
it can be `return process.system_output(cmd_governor_list).split()` just want to eliminate extra variable

PavithraDP replied:
done

PraveenPenguin commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        cmd3 = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_cur_freq" % (self.i)
+        min = process.system_output(cmd1).strip()
+        max = process.system_output(cmd2).strip()
+        cur = process.system_output(cmd3).strip()
```
here  also

PraveenPenguin commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        cmd1 = "cpupower -c %s frequency-info -f" % (self.i)
+        output1 = process.system_output(cmd1).splitlines()[1]
+        freq = output1.split()[3]
+        return freq
```
can this be  `return process.system_output(cmd1).splitlines()[1].split()[3]`

PavithraDP replied:
done

PraveenPenguin commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        value2 = []
+        value2 = value.split(' ')
+        rand = random.choice(value2)
+        return rand
```
here also `return  random.choice(process.system_output(cmd).split(' '))` please

PavithraDP replied:
done

PraveenPenguin commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        Validate Userspace Governor
+        """
+        self.set_governor(governor)
+        for j in range(3):
```
please use a meaningful variable name as `j `not looks good

PavithraDP replied:
done

PraveenPenguin commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        self.set_governor(governor)
+        for j in range(3):
+            rand = self.get_random_freq()
+            self.set_freq_val(rand)
```
here also it can be `self.set_freq_val(self.get_random_freq())`

PavithraDP replied:
done

balamuruhans commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        """
+        cmd1 = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_min_freq" % (self.i)
+        cmd2 = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_max_freq" % (self.i)
+        cmd3 = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_cur_freq" % (self.i)
```
To avoid redundancy, you can have a method to form it. As it is almost similar every where,
```

def cpu_freq_path(self, cpu_num, path=None):
    cmd = "cat /sys/devices/system/cpu/cpu%s/cpufreq" % cpu_num
    if path:
        cmd = os.path.join(cmd, path)
    return cmd
```

PraveenPenguin replied:
i think it can be like 

```
def cpu_freq_path(self, file, cpu_num, path=None):
    filename = "/sys/devices/system/cpu/cpu%s/cpufreq/%s" % (cpu_num, file)
    return  open(filename, 'r').readline().strip('\n')
```

function calling can be 

`min = self.cpu_freq_path('scaling_min_freq',self.i)`

it applicable for every where @balamuruhans what you say ?

PavithraDP replied:
done

balamuruhans replied:
@PraveenPenguin I agree, then path=None is not required.

balamuruhans commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+
+    def get_freeq(self):
+        """
+        Get the cur,min,max freequency on the system
```
%s/freequency/frequency

PavithraDP replied:
done

balamuruhans commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        governors = output.split()
+        return governors
+
+    def get_freeq(self):
```
def get_freq(self):

PavithraDP replied:
done

balamuruhans commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+                self.log.info("Dynamic governors,need manual verification")
+        self.log.info("Set the final values on the system")
+        self.final_freq(initial_freq, initial_governor)
+        if self.nfail != 0:
```
`if self.nfail:` will be sufficient

PavithraDP replied:
done

balamuruhans commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+    def set_governor(self, governor):
+        """
+        Set the governor specified in the argument
+        """
```
docstring for the param governer

PavithraDP replied:
done

balamuruhans commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+
+    def get_random_freq(self):
+        """
+        Get random freequency from list
```
typo for frequency

PavithraDP replied:
done

balamuruhans commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+
+    def set_freq_val(self, freq):
+        """
+        Set the freequency value specified in argument
```
typo frequency

balamuruhans commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+    def final_freq(self, initial_freq, initial_governor):
+        """
+        Set the system with Intial values
+        """
```
docstring for initial_freq, initial_governor

PavithraDP replied:
done

sathnaga commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        """
+        Validate Performance Governor
+        """
+        self.set_governor(governor)
```
return success from set_governor() then check here and if then proceed for checking freq values

PavithraDP replied:
done

sathnaga commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+            self.log.info("%s governor working as expected" % governor)
+        else:
+            self.log.error("%s governor not working as expected" % governor)
+            self.nfail += 1
```
Have a template for values to be checked for different governors in a dict  and use common function(check_governor) for all governors that would help in future incase of additional governors to be added.

sathnaga commented on cpupower/cpupower.py:
```diff
@@ -0,0 +1,181 @@
+        cmd1 = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_governor" % (self.i)
+        initial_governor = process.system_output(cmd1)
+        return (initial_freq, initial_governor)
+
```
this function can be clubbed with get_freq values as @balamuruhans suggested

PavithraDP commented:
Have done the following:
1) Incorporated the changes
2) Combined 2 functions (get_freq and get_initial_	freq) - Made it into single function get_initial_values
3) Added 2 new functions check_governor and cpu_freq_path for reducing redundancy of fetching values from files(scaling_cur_min,scaling_cur_max).

PraveenPenguin commented:
@PavithraDP job log please

PavithraDP commented:
Job Log
========
[job.log.tar.gz](https://github.com/avocado-framework-tests/avocado-misc-tests/files/689677/job.log.tar.gz)

PraveenPenguin commented:
@PavithraDP sorry just i noticed that you added this test case in new category as `cpupower`  i think it can be move in `cpu`  so moving this under `cpu`

PraveenPenguin commented:
@balamuruhans @sathnaga @apahim thanks

PavithraDP commented:
@PraveenPenguin @sathnaga @apahim @balamuruhans  Thanks for the review.
@PraveenPenguin Thanks for merging and closing this PR
